### ISSUE TITLE
Callable alias with `execx` can provoke `Bad file descriptor` exception in the internal or external code

### ISSUE THREAD

**Issue by anki-code** [xonsh/xonsh#5631]
### How it looks for a user

The exceptions below are not the root of the issue. It's just consequences of closing the descriptor somewhere in upstream or in parallel thread during execution of callable alias.

Where I can catch this exception:
* In case of using ExecAlias e.g. `aliases['a'] = "echo @(1)"`
* In case of using `execx` in callable alias e.g. `aliases['a'] = lambda: execx('echo 1')`
* I catch this during `conda` and `mamba` hooks usage.

How exception looks in ptk:
```xsh
xonsh --no-rc
aliases['t'] = 'echo 1 && echo 2'  # or just 'echo @(1)'
aliases['t']
# ExecAlias
t
t
t
# ...
```
```xsh
  File "/Users/pc/git/xonsh/xonsh/main.py", line 512, in main
    sys.exit(main_xonsh(args))
             ^^^^^^^^^^^^^^^^
  File "/Users/pc/git/xonsh/xonsh/main.py", line 558, in main_xonsh
    shell.shell.cmdloop()
  File "/Users/pc/git/xonsh/xonsh/shells/ptk_shell/__init__.py", line 415, in cmdloop
    line = self.singleline(auto_suggest=auto_suggest)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pc/git/xonsh/xonsh/shells/ptk_shell/__init__.py", line 383, in singleline
    line = self.prompter.prompt(**prompt_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/site-packages/prompt_toolkit/shortcuts/prompt.py", line 1024, in prompt
    return self.app.run(
           ^^^^^^^^^^^^^
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/site-packages/prompt_toolkit/application/application.py", line 973, in run
    return loop.run_until_complete(coro)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
    self.run_forever()
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
    self._run_once()
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/asyncio/base_events.py", line 1949, in _run_once
    event_list = self._selector.select(timeout)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/selectors.py", line 566, in select
    kev_list = self._selector.control(None, max_ev, timeout)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 9] Bad file descriptor
```
How exception looks in readline:
```xsh
xonsh --no-rc -st readline
aliases['t'] = 'echo 1 && echo 2' # ExecAlias
t
t
t
# ...
```
```xsh
Exception in thread Thread-52:
Traceback (most recent call last):
  File "/Users/pc/.local/mamba-envs/Users_pc_git_xonsh/lib/python3.12/threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "/Users/pc/git/xonsh/xonsh/procs/proxies.py", line 444, in run
    open(self.c2pwrite, "wb", -1), encoding=enc, errors=err
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 9] Bad file descriptor

```

### Workaround

https://github.com/xonsh/xonsh/issues/5631#issuecomment-2250226809

### Minimal PoC

https://github.com/xonsh/xonsh/issues/5631#issuecomment-2250423483

### Probable cause

After massive refactoring in #3897 we have situations when the same pipe file descriptor can be opened/closed many times. The details are below in the thread.

Cc #5241 #5437

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by jnoortheen** [comment]
what about using named pipes and tmpfiles ?

**Comment by gforsyth** [comment]
I can't say that I remember reviewing that PR, although I obviously did.

I haven't been in these guts for a while, it'll take some time to get back into it.

I agree that it's a race-condition around the double-open.

Not sure if named-pipes are an option on windows (they have them but I don't remember if `subprocess` can use them or if we'll end up needing to mess with `win32api` or whatever that thing is called)

**Comment by jaraco** [comment]
I don't have much insight to add. I've not done a lot of work with pipes. The findings in https://github.com/xonsh/xonsh/issues/5631#issuecomment-2252067259 are surprising. I don't yet even understand the difference between Case 1 and Case 2. Oh, maybe I do. I was interpreting "after opening multiple times" as "after opening any number of times", but I see now it's just the two times that are allowed.

That's strange that it fails only on the third operation (for any given `pipe()`). I wonder - do we have documentation on how these pipes are specified to be used?

**Comment by jaraco** [comment]
The [Python docs](https://docs.python.org/3/library/os.html#os.pipe) don't have much to say.

**Comment by jaraco** [comment]
I think I've figured out why the `open` fails on the third call (and not the second).

After the first open, an IOWrapper is created but not bound to anything. During the second open, the first IOWrapper is garbage collected and closes the fd. Then, during the third open, the fd is bad.

The same behavior can be illustrated to fail after the second open with:

```
>>> r, W = os.pipe()
>>> open(W, 'w'); open(0, 'r'); open(W, 'w')
<_io.TextIOWrapper name=7 mode='w' encoding='UTF-8'>
<_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 9] Bad file descriptor
```

It seems that the intervening operation must be a call to `open`, however. Using a `time.sleep(.1)` between the calls does not allow the fd to become bad. It seems that something must be calling into the fd handling logic.

Interestingly, keeping handles to the opened objects prevents them from being closed and becoming bad:

```
>>> r, W = os.pipe()
>>> a = open(W, 'w')
>>> b = open(W, 'w')
>>> c = open(W, 'w')
>>> 
```

Maybe all that's needed is to keep a handle on the opened pipes and only open them once?

**Comment by anki-code** [comment]
I've asked Sonnet 4.6 (extended):
```
Read this ticket in detail - https://github.com/xonsh/xonsh/issues/5631
Examine the xonsh/procs/*.py code in detail.
Ignore anything related to the Windows logic and focus on the core logic (default POSIX Linux).
Don't rush your answer.
What issues could be in the callable alias code (ProcProxy, ProcProxyThreadable) 
that lead to the issues mentioned in the ticket?
What can be done?
```

The result:

------

## How callable alias I/O is set up

When xonsh runs a callable alias (or `ExecAlias`), the pipeline is prepared in two separate phases:

**`specs.py` — pipe creation:**
```python
# For each `|` in a pipeline like `cmd | alias_func`:
r, w = os.pipe()
specs[i].stdout   = w          # raw int FD  or  safe_open(w, "wb")
specs[i+1].stdin  = r          # raw int FD  or  safe_open(r, "rb")

# For the last command:
r, w = os.pipe()               # or openpty()
last.stdout          = safe_open(w, "w")   # TextIOWrapper around FD w
last.captured_stdout = safe_open(r, "r")   # TextIOWrapper around FD r
```

**`ProcProxyThread.__init__()` — stores raw FD numbers:**
```python
if isinstance(stdin, int):
    self.stdin = io.open(stdin, "wb", -1)   # wraps the raw int FD
else:
    self.c2pwrite = stdout.fileno()          # extracts the raw int
    self.errwrite  = stderr.fileno()         # extracts the raw int
```

**`ProcProxyThread.run()` — opens the FD again in a new thread:**
```python
sp_stdout = io.TextIOWrapper(
    io.open(self.c2pwrite, "wb", -1),        # ← opens AGAIN
    encoding=enc, errors=err
)
```

---

## Problem 1: Double-wrapping — two Python objects sharing one OS FD

This is the central issue.

`specs.py` creates a `TextIOWrapper` around FD 9:
```python
last.stdout = safe_open(w, "w")   # TextIOWrapper name=9
```

`ProcProxyThread.__init__()` receives that object and does:
```python
self.c2pwrite = stdout.fileno()   # = 9
```

Then `run()` creates a **second** `TextIOWrapper` around the same FD 9:
```python
io.open(self.c2pwrite, "wb", -1)  # another Python file object on FD 9
```

Now two Python objects own the same OS FD. Whoever closes first destroys the descriptor for the other. When `CommandPipeline._end()` closes `last.stdout`, FD 9 is released. The `ProcProxyThread` thread then gets `EBADF` on its next write attempt.

---

## Problem 2: TOCTOU race — FD stored as int, opened later in the thread

The FD is alive in `__init__()`. By the time the thread actually calls `run()`, the FD may already be closed:

```
Main thread:                          ProcProxyThread:
__init__: self.c2pwrite = 9
start()
CommandPipeline cleans up pipeline
  → closes file objects
  → close(fd=9)  ← FD 9 is now free
                                      run() is called:
                                        io.open(self.c2pwrite=9)
                                        → OSError: [Errno 9] EBADF
```

This is a classic TOCTOU race between the spawned thread and the main thread.

---

## Problem 3: FD reuse — OS recycles the closed FD number

After FD 9 is closed, the OS gives that number to the next `open()` call — which may be a `kqueue` / `epoll` / `selector` fd created internally by `prompt_toolkit`.

Now `ProcProxyThread.run()` calls `io.open(9, "wb")` and accidentally opens the **ptk selector fd** as a write pipe. When the thread writes to it, ptk's event loop is corrupted. When ptk calls `selector.select()` on that fd, it gets `OSError: [Errno 9] Bad file descriptor`.

This explains exactly the traceback from the issue:
```
File ".../selectors.py", line 566, in select
    kev_list = self._selector.control(None, max_ev, timeout)
OSError: [Errno 9] Bad file descriptor
```

---

## Problem 4: `execx()` inside callable alias creates a new pipeline

`ExecAlias` (`aliases['t'] = 'echo 1 && echo 2'`) calls `execx()` **from inside** a running `ProcProxyThread.run()`.

`execx()` builds a brand new sub-pipeline with fresh `os.pipe()` calls. Those new calls may receive FD numbers that were just freed from the outer pipeline — causing conflicts:

```
Outer pipeline:  stdin=FD3, stdout=FD7, stderr=FD9
execx() builds sub-pipeline:
  → os.pipe() → (FD3, FD7)    # OS recycled these numbers!
  → subprocess.Popen for echo  → closes write-end FD7
  → FD7 was ProcProxyThread's stdout → now it's gone
```

Additionally, `execx()` is unaware of the current `ProcProxyThread`'s stdin/stdout/stderr and operates on `sys.stdin/stdout/stderr` directly, potentially bypassing the pipe setup entirely.

---

## Problem 5: No clear FD ownership

On POSIX systems each FD should have **exactly one owner** responsible for closing it. In the current code there is no clear protocol:

- `specs.py` creates FDs and wraps them in file objects
- `ProcProxyThread` extracts `fileno()` and wraps them again
- `CommandPipeline._end()` closes the file objects from `specs`
- `ProcProxyThread.__del__` / `_clean_up()` tries to close its own objects
- Result: double-close or close-while-still-in-use

---

## Solutions

### Solution 1 (core fix): `os.dup()` — give `ProcProxyThread` its own FD copy

The key idea: whenever `ProcProxyThread` receives an FD, immediately **duplicate** it:

```python
# In __init__:
if isinstance(stdin, int):
    self._owned_stdin_fd = os.dup(stdin)
elif stdin is not None:
    self._owned_stdin_fd = os.dup(stdin.fileno())
else:
    self._owned_stdin_fd = None

# Same for stdout, stderr
```

Benefits:
- `specs.py` can safely close its own file objects without affecting `ProcProxyThread`
- `ProcProxyThread` exclusively owns its FDs and is solely responsible for closing them
- Each Python file object wraps a unique FD — no double-wrap

---

### Solution 2: Open the file object in `__init__`, not in `run()`

Instead of storing the raw int and opening it later:
```python
self.c2pwrite = stdout.fileno()   # store int, open later in run()
```

Open it immediately in `__init__()` using `dup()`:
```python
self.proc_stdout = io.TextIOWrapper(
    io.open(os.dup(stdout.fileno()), "wb", -1),
    encoding=enc, errors=err
)
```

This closes the entire TOCTOU window between `__init__` and `run()`.

---

### Solution 3: Explicit transfer of FD ownership in `specs.py`

`specs.py` should pass raw FDs with clear "ownership transfer" semantics instead of pre-wrapping them:

```python
# In specs.py when building ProcProxyThread:
stdout_fd = w   # raw int from os.pipe()
kwargs['stdout'] = stdout_fd   # pass the raw int

# After ProcProxyThread.__init__ has dup()-ed it, close the parent's copy:
os.close(stdout_fd)
```

This is standard POSIX practice: parent passes FD to child via `dup()`, then closes its own copy. Only one entity holds each FD at any given time.

---

### Solution 4: Redirect `sys.stdin/stdout/stderr` inside `ExecAlias`

When a callable alias calls `execx()`, it must run in the correct I/O context:

```python
import sys

old_stdout, old_stderr = sys.stdout, sys.stderr
sys.stdout = self.proc_stdout
sys.stderr = self.proc_stderr
try:
    execx(cmd_string)
finally:
    sys.stdout = old_stdout
    sys.stderr = old_stderr
```

Without this, `execx()` writes directly to the shell's `sys.stdout` (i.e. the ptk terminal), bypassing the pipe pipeline entirely and causing output to appear out-of-order or in the wrong place.

---

### Solution 5 (short-term workaround): Graceful `EBADF` handling in `run()`

As a temporary band-aid, catch `OSError: [Errno 9]` before it corrupts ptk:

```python
import errno

try:
    sp_stdout = io.TextIOWrapper(io.open(self.c2pwrite, "wb", -1), ...)
except OSError as e:
    if e.errno == errno.EBADF:
        # FD was closed or recycled — this is the race condition
        return   # or log a warning
    raise
```

This prevents ptk corruption but does not fix the underlying race. It should be paired with one of the ownership solutions above.

---

## Summary table

| Problem | Root cause | Fix |
|---|---|---|
| Double-wrap of the same FD | `specs.py` + `ProcProxyThread` both create file objects on the same raw FD | `os.dup()` in `__init__()` |
| TOCTOU: FD alive in `__init__`, dead in `run()` | Raw int stored, file opened only later in the thread | Open immediately in `__init__()` via `dup()` |
| FD reuse corrupts ptk selector | Freed FD number recycled by ptk's `kqueue`/`epoll` fd | `os.dup()` isolates from recycling |
| `execx()` in ExecAlias bypasses pipe I/O | Sub-pipeline created without the outer thread's I/O context | Redirect `sys.stdout/stderr` around `execx()` call |
| No clear FD ownership | Multiple code paths close the same FDs | Ownership transfer protocol + `dup()` |

The fix for all of the above boils down to a single principle: **each OS file descriptor must have exactly one Python owner, established at the moment the FD is received, via `os.dup()`**.

------

cc #3897

**Comment by anki-code** [comment]
I've tried to apply "Solution 1 (core fix): os.dup() — give ProcProxyThread its own FD copy" with LLM and it's the right direction that fixes the general issue but further testing shows that it's not complete solution and may be we need to change logic more deeply. LLM can't give me the complete fix after many tries and fails on this test:

```xsh
aliases['t'] = 'echo 1 && echo 2'  # or aliases['t'] = 'echo @(1)'   # ExecAlias 
t
t
t
t
t | t
t | t
t | t
t | t
t | t
t | t
# ValueError('I/O operation on closed file.')
# echo: fflush: Input/output error
```
Because we see that one command affects another I think something is wrong with logic between CommandPipeline, Specs, ProcProxy, etc. Command pipeline open descriptors for ExecAlias but doesn't close it properly because of race condition or deadlock logic.

**Comment by anki-code** [comment]
New from Sonet:

Root Cause Analysis: `sys.stderr` Gets Closed by `ProcProxyThread`

## How stderr is configured in `specs.py`

Depending on the capture mode, stderr is set up differently:

```python
# captured == "object"  (e.g. !(alias)):
r, w = os.pipe()
last.stderr          = safe_open(w, "w")   # TextIOWrapper(FD=w_err)
last.captured_stderr = safe_open(r, "r")

# captured == "stdout":
pass   # stderr is not touched → remains None

# captured == "hiddenobject" (normal interactive alias run):
pass   # stderr is also None in most cases
```

**Key point:** in the most common case of an interactive alias run, `stderr=None` is passed into `ProcProxyThread`.

---

## `ProcProxyThread.__init__()` — what happens with `None`

```python
def __init__(self, f, args, stdin=None, stdout=None, stderr=None, ...):
    ...
    if stderr is None:
        self.errwrite = 2        # ← FD 2, i.e. the shell's own sys.stderr!
    elif isinstance(stderr, int):
        self.errwrite = stderr
    else:
        self.errwrite = stderr.fileno()
```

When `stderr=None` → `self.errwrite = 2`. This is not a copy, not a `dup()` — it is literally **the global FD=2 of the entire process**.

---

## `ProcProxyThread.run()` — opening FD=2 as a new Python object

```python
enc = env.get("XONSH_ENCODING")
err = env.get("XONSH_ENCODING_ERRORS")

sp_stderr = io.TextIOWrapper(
    io.open(self.errwrite, "wb", -1),   # ← io.open(2, "wb", -1)
    encoding=enc, errors=err
)
```

`io.open(2, "wb", -1)` creates a **new Python `BufferedWriter` object that takes ownership of FD=2**.

---

## End of `run()` — closing `sp_stderr`

At the end of `run()` (explicitly or via GC):

```python
sp_stderr.close()   # → flush() + close() on BufferedWriter
                    # → os.close(2)  ← FD 2 is now DEAD
```

**After this point FD=2 for the entire process is gone.**

---

## What breaks after FD=2 is closed

### The shell's `sys.stderr` is permanently broken

```python
sys.stderr   # TextIOWrapper wrapping FD=2
             # FD=2 is closed → any write raises:
             # ValueError: I/O operation on closed file
             # or OSError: [Errno 9] Bad file descriptor
```

### `prompt_toolkit` crashes

ptk writes to stderr when rendering the prompt, reporting errors, and calling `print_formatted_text`. All of that breaks:

```
File ".../prompt_toolkit/output/vt100.py"
    self._output.write(...)
ValueError: I/O operation on closed file.
```

### The next callable alias fails on startup

The next `ProcProxyThread.__init__()` once again sets `self.errwrite = 2` (because `stderr=None`). Then `run()` calls `io.open(2, "wb", -1)` — but FD=2 is already closed → `OSError: [Errno 9] EBADF` right at thread startup.

---

## Full lifetime map of FD=2

```
Moment           Who holds FD=2
────────────────────────────────────────────────────────────────
Shell start      sys.stderr = TextIOWrapper(FD=2)
                     │
                     │  (forever — this is the global stderr)
                     │
Alias launched   ProcProxyThread.__init__:
                     self.errwrite = 2    ← just stores the int, no dup
                     │
                 thread.start()
                     │
run() starts     io.open(2, "wb") → sp_stderr   ← second object on FD=2
                     │
                 alias function is called ...
                     │
run() finishes   sp_stderr.close()
                     → os.close(2)   ← FD=2 IS NOW DEAD
                     │
                     ▼
                 sys.stderr.write(...)  → EBADF / closed file error
```

---

## Comparison with the stdout pipe problem

| | stdout | stderr |
|---|---|---|
| FD source in specs | fresh `os.pipe()` FD, unique per command | often `None` → falls back to FD=2 |
| Uniqueness | unique to each pipeline | **global to the entire process** |
| Consequence of close | EOF never arrives, pipeline hangs | `sys.stderr` of the shell is broken forever |
| Recovery | happens automatically on next pipe | none — requires `os.dup2()` or shell restart |

The stderr problem is **worse** than the stdout problem. Closing one pipeline's write-end only affects that pipeline. Closing FD=2 breaks **the entire shell**.

---

## Root cause: why does `stderr=None` lead to `io.open(2, ...)`?

This is a design decision: a callable alias is "allowed" to write directly to the shell's real stderr. But the implementation does this unsafely — by opening a **new Python object on FD=2** instead of simply passing `sys.stderr` through.

The correct approaches would be:

**Option A — pass `sys.stderr` as-is, without wrapping:**
```python
# In run(): do not create a new object, do not close it
sp_stderr = sys.stderr
```

**Option B — if a separate object is needed, use `os.dup()`:**
```python
sp_stderr = io.TextIOWrapper(
    io.open(os.dup(2), "wb", -1),   # dup() → e.g. FD=7 (new, independent)
    encoding=enc, errors=err
)
# now close() will close FD=7, and FD=2 stays alive
```

Both options ensure that `sp_stderr.close()` at the end of `run()` **never touches FD=2**, so `sys.stderr` remains usable for the rest of the shell's lifetime.

**Comment by anki-code** [comment]
```xsh
$XONSH_TRACE_SUBPROC=3
@aliases.register
def _x1(args=None, stdin=None, stdout=None):
   echo @(123)
   execx('echo @(321)')
   print("111")

@aliases.register
def _x2(args=None, stdin=None, stdout=None):
   print("222")

x1 | x2

Trace run_subproc({'cmds': (['x1'], '|', ['x2']), 'captured': 'hiddenobject'})
0: {'cmd': ['x1'], 'cls': 'xonsh.procs.proxies.ProcProxyThread', 'alias_name': 'x1', 'alias': FuncAlias({'name': 'x1', 'func': '_x1', 'return_what': 'result'}), 'threadable': True, 'background': False, 'stdout': 4, 'captured': 'hiddenobject'}
1: {'cmd': ['x2'], 'cls': 'xonsh.procs.proxies.ProcProxyThread', 'alias_name': 'x2', 'alias': FuncAlias({'name': 'x2', 'func': '_x2', 'return_what': 'result'}), 'threadable': True, 'background': False, 'stdin': 3, 'stdout': <_io.TextIOWrapper name=6 mode='w' encoding='UTF-8'>, 'stderr': <_io.TextIOWrapper name=8 mode='w' encoding='UTF-8'>, 'captured': 'hiddenobject', 'captured_stdout': <_io.TextIOWrapper name=5 mode='r' encoding='UTF-8'>, 'captured_stderr': <_io.TextIOWrapper name=7 mode='r' encoding='UTF-8'>}
222
Trace run_subproc({'cmds': (['echo', '123'],), 'captured': 'hiddenobject'})
0: {'cmd': ['echo', '123'], 'cls': 'xonsh.procs.posix.PopenThread', 'binary_loc': '/bin/echo', 'threadable': True, 'background': False, 'stdout': <_io.TextIOWrapper name=6 mode='w' encoding='UTF-8'>, 'stderr': <_io.TextIOWrapper name=7 mode='w' encoding='UTF-8'>, 'captured': 'hiddenobject', 'captured_stdout': <_io.TextIOWrapper name=5 mode='r' encoding='UTF-8'>, 'captured_stderr': <_io.TextIOWrapper name=3 mode='r' encoding='UTF-8'>}
123
Trace run_subproc({'cmds': (['echo', '321'],), 'captured': 'hiddenobject'})
0: {'cmd': ['echo', '321'], 'cls': 'xonsh.procs.posix.PopenThread', 'binary_loc': '/bin/echo', 'threadable': True, 'background': False, 'stdout': <_io.TextIOWrapper name=5 mode='w' encoding='UTF-8'>, 'stderr': <_io.TextIOWrapper name=7 mode='w' encoding='UTF-8'>, 'captured': 'hiddenobject', 'captured_stdout': <_io.TextIOWrapper name=3 mode='r' encoding='UTF-8'>, 'captured_stderr': <_io.TextIOWrapper name=6 mode='r' encoding='UTF-8'>}
321
111
```

LLM Opus:

Looking at the FD numbers across three pipelines from the trace:

## Outer pipeline `x1 | x2`

```
spec[0] x1: stdout=4
spec[1] x2: stdin=3, stdout=6(w), stderr=8(w), captured_stdout=5(r), captured_stderr=7(r)
```

FDs occupied by the outer pipeline: **3, 4, 5, 6, 7, 8**.

## Inner `echo 123` (inside x1)

```
stdout=6(w), stderr=7(w), captured_stdout=5(r), captured_stderr=3(r)
```

## Inner `echo 321` (inside x1, via execx)

```
stdout=5(w), stderr=7(w), captured_stdout=3(r), captured_stderr=6(r)
```

---

## Collisions — at a glance

| FD | Outer pipeline | echo 123 | echo 321 |
|----|----------------|----------|----------|
| **3** | x2 stdin (pipe_r) | captured_stderr(r) | captured_stdout(r) |
| **5** | x2 captured_stdout(r) | captured_stdout(r) | stdout(w) |
| **6** | x2 stdout(w) | stdout(w) | captured_stderr(r) |
| **7** | x2 captured_stderr(r) | stderr(w) | stderr(w) |

**Every** FD of the inner pipelines collides with an FD from the outer one. This is direct evidence of the reuse race.

## What exactly happened

**1. x2 finished before x1.** Visible from the output order: `222` appeared before `123`. Thread x2 executed `print("222")`, completed `run()`, and closed its FDs: stdin(3), stdout(6). Possibly the pipeline also closed 5, 7 on its side. All these numbers became free.

**2. `echo 123` got reused numbers.** `os.pipe()` inside the inner pipeline returned (5,6) and (3,7) — exactly the FDs freed by the outer pipeline.

**3. `echo 321` got the same numbers reshuffled again.** After cleanup of `echo 123`, FDs 3,5,6,7 were freed — and `echo 321` got them again, but in a different read/write combination.

## Most dangerous collisions

**FD 3** — this is pipe_r between x1→x2 in the outer pipeline. The trace shows that the inner `echo 123` got `captured_stderr=3(r)`. When the inner pipeline cleans up and closes its captured_stderr → `close(3)`, it **destroys** the pipe between x1 and x2. If the outer `_close_prev_procs` then tries to `close(3)`, that number may already belong to someone else (PTK event loop).

**FD 5** — the outer `captured_stdout` for reading x2's output. The inner `echo 123` also has `captured_stdout=5(r)`. Closing 5 in the inner pipeline kills the outer pipeline's ability to read x2's output (if it hasn't been read yet).

**FD 6 and 7** — same situation: r/w ends of pipes are shared between the outer and inner pipelines.

## Why this particular run didn't crash

Got lucky with timing: x2 **already finished** and its output (`222`) was already read before the inner pipeline reused the FDs. But this is a pure race — with different thread scheduling (x1 starts execx before x2 finishes) it will crash. This is exactly why the bug reproduces "almost immediately" but not with 100% consistency.

**Comment by anki-code** [comment]
LLM's thinking in the right direction but can't form complete solution. To fix this we need to go thru the full cycle:
1. creating CommandPipeline
2. Setting stdin/out/err fds in Specs
3. Creating pipes and fds
4. Waiting for outputs
5. Closing specs/threads fds
6. Closing specs/threads fds in Specs
7. Closing specs/threads fds in CommandPipeline (_end function)

I think there are two issues:
1. Lack of right fd management for ProcProxyThread by using os.dup OR using just one owner of fd.
2. The pipeline from p1 to p7 has issue in logic in case of using `execx` inside callable alias.

LLM can't bring all of these two things together and make complete fix so it's needed to human brain to start from tracing p2 and apply p1.

Overall it's a good issue and it will be highly appreciate to solve it.

**Comment by anki-code** [comment]
After additional testing I think that there is another source of issue (I found this in another tickets before but now I see it again). May be the issue is in the case when previous process in the pipe closed before the last process (because of error or just because there is no making the output). 

Here we have no proper waiting for previous processes in case of proc (last proc) was ended:

https://github.com/xonsh/xonsh/blob/b65a981896ac4c7956700a3dfdd68ba5d5f1a68c/xonsh/procs/pipelines.py#L303

Example - this is working perfect because a1 produces the output and a2 consume the output and they are working in sync:
```xsh
def _a(args, stdin, stdout):
    if stdin is None:
        print('first', file=stdout, flush=True)
    else:
        for line in stdin.readlines():
            print(line, file=stdout, flush=True)
        
aliases['a1'] = _a
aliases['a2'] = _a

for i in range(1,50):
    a1 | a2
# first
# first
# ...
# Everything is working
```

But in this example - e1/e2 can be closed before e2/e1 because there is no connection between e1 and e2 - they just print 1 and exit:
```xsh
def _e(args, stdin, stdout):
    print(1)

aliases['e1'] = _e
aliases['e2'] = _e

for i in range(1,50):
    e1 | e2
# 1
# 1
#  File "/Users/pc/.local/xonsh-env/lib/python3.12/site-packages/xonsh/procs/proxies.py", line 444, in run
#    open(self.c2pwrite, "wb", -1), encoding=enc, errors=err
#    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# OSError: [Errno 9] Bad file descriptor
```

So this means that we have fragile implementation of the part that is waiting for process ending in the pipe. And this is why any error in callable alias can interrupt the execution and then follow to I/O errors in the current session in entire shell.

UPD: the issue appears when e2 (the last one) closed before previous processes.

But maybe this is not related to using execx and this is separate issue.

(The LLMs hypotesis about fd is near execx issue and not exactly this case with last proc.)

**Comment by anki-code** [comment]
I just want to leave this here:
```xsh
xonsh --no-rc --no-env

print('base')
print(@.imp.sys.stdout)
print(@.imp.sys.stderr)

print('base-execx')
execx("print(@.imp.sys.stdout)")
execx("print(@.imp.sys.stderr)")


@aliases.register
def _a(stdout, stderr):
    print('alias')
    print(stdout)
    print(stderr)
    
    print('alias-execx')
    execx("print(@.imp.sys.stdout)")
    execx("print(@.imp.sys.stderr)")

a

print('=== Captured stdout alias ===========')
print($(a))

print('=== Captured object alias ===========')
print(!(a).out)
```
```xsh
base
    <xonsh.shells.base_shell._TeeStd object at 0x1012c77f0>
    <xonsh.shells.base_shell._TeeStd object at 0x1029703a0>
base-execx
    <xonsh.shells.base_shell._TeeStd object at 0x1012c77f0>
    <xonsh.shells.base_shell._TeeStd object at 0x1029703a0>
alias
    <_io.TextIOWrapper name=4 encoding='utf-8'>
    <_io.TextIOWrapper name=6 encoding='utf-8'>
alias-execx
    <xonsh.procs.proxies.FileThreadDispatcher object at 0x100c46a50>
    <xonsh.procs.proxies.FileThreadDispatcher object at 0x101fe7290>

=== Captured stdout alias ===========
alias
    <_io.TextIOWrapper name=4 encoding='utf-8'>
    <xonsh.shells.base_shell._TeeStd object at 0x1029703a0>
alias-execx
    <xonsh.procs.proxies.FileThreadDispatcher object at 0x100c46a50>
    <xonsh.procs.proxies.FileThreadDispatcher object at 0x101fe7290>

=== Captured object alias ===========
alias
    <_io.TextIOWrapper name=4 encoding='utf-8'>
    <_io.TextIOWrapper name=6 encoding='utf-8'>
alias-execx
    <xonsh.procs.proxies.FileThreadDispatcher object at 0x100c46a50>
    <xonsh.procs.proxies.FileThreadDispatcher object at 0x101fe7290>
```
```xsh
Python 3.11.14 | packaged by conda-forge | (main, Jan 27 2026, 00:01:01) [Clang 19.1.7 ] on darwin
>>> import sys
>>> sys.stdout
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
>>> sys.stderr
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
>>> sys.stdout.fileno()
1
>>> sys.stderr.fileno()
2
>>>
```

**Comment by anki-code** [comment]
My best try today was in #6126. Amen.

**Comment by anki-code** [comment]
For researchers: the execution workflow (based on b65a981896ac4c7956700a3dfdd68ba5d5f1a68c):
`run_subproc` 
-> `cmds_to_specs: SubprocSpec.build() + pipe` 
-> `_update_last_spec(_last_spec_update_captured(_make_last_spec_captured()))` 
-> `_run_specs` 
-> `_run_command_pipeline` 
-> `CommandPipeline(specs)` 
-> `for spec in specs: spec.run()` 
-> `Run self.cls as ProcProxyThread (or _run_binary)` 
-> `ProcProxyThread __init__ and run()` (in thread)
-> `Run callable alias with run_with_partial_args` 
-> `safe_fdclose stdout stderr` (here is the issue about it's not really safe)
-> `CommandPipeline end() -> _end()` 
-> `tee_stdout()` -> `iterraw()` -> `proc.poll()` (here is the issue about no waiting for all commands)
-> `read stdout` -> `close` -> `close` -> `close` -> `end`

**Comment by anki-code** [comment]
Fixed in #6159
